File: rag/vectorstore.py
"""ChromaDB-backed vector store for document chunks."""
from __future__ import annotations

import hashlib
from dataclasses import dataclass
from pathlib import Path
from typing import Callable, List, Optional

import chromadb
from chromadb.config import Settings

from . import embeddings


PERSIST_DIR = Path("chroma_db")


@dataclass
class Retrieved:
    text: str
    score: float  # similarity (higher = better)
    metadata: dict


class VectorStore:
    def __init__(self, collection_name: str, persist_dir: Path = PERSIST_DIR):
        persist_dir.mkdir(parents=True, exist_ok=True)
        self.client = chromadb.PersistentClient(
            path=str(persist_dir),
            settings=Settings(anonymized_telemetry=False),
        )
        # cosine works well with normalized text embeddings.
        self.collection = self.client.get_or_create_collection(
            name=collection_name,
            metadata={"hnsw:space": "cosine"},
        )

    @staticmethod
    def make_collection_name(filename: str, content: bytes) -> str:
        """Stable per-document collection name so re-uploads reuse the index."""
        digest = hashlib.sha1(content).hexdigest()[:12]
        # Chroma collection names: 3-63 chars, alnum + _ - . only.
        safe = "".join(c if c.isalnum() else "_" for c in filename)[:40]
        return f"doc_{safe}_{digest}"

    def has_data(self) -> bool:
        try:
            return self.collection.count() > 0
        except Exception:
            return False

    def add_chunks(
        self,
        chunks: List[str],
        metadatas: Optional[List[dict]] = None,
        on_progress: Optional[Callable[[int, int], None]] = None,
    ) -> None:
        if not chunks:
            return
        vectors = embeddings.embed_documents(chunks, on_progress=on_progress)
        ids = [f"chunk_{i}" for i in range(len(chunks))]
        metas = metadatas or [{"chunk_index": i} for i in range(len(chunks))]
        self.collection.add(
            ids=ids,
            documents=chunks,
            embeddings=vectors,
            metadatas=metas,
        )

    def query(self, question: str, k: int = 4) -> List[Retrieved]:
        query_vec = embeddings.embed_query(question)
        results = self.collection.query(
            query_embeddings=[query_vec],
            n_results=k,
        )
        docs = results.get("documents", [[]])[0]
        metas = results.get("metadatas", [[]])[0]
        # Chroma returns cosine *distance*; convert to similarity (1 - distance).
        dists = results.get("distances", [[]])[0]

        retrieved = []
        for text, meta, dist in zip(docs, metas, dists):
            retrieved.append(
                Retrieved(text=text, score=1.0 - float(dist), metadata=meta or {})
            )
        return retrieved

File: src/vectorstore.js
/**
 * ChromaDB-backed vector store for document chunks.
 *
 * The JavaScript Chroma client talks to a running Chroma server (unlike the
 * Python embedded PersistentClient). Start one first:
 *
 *     pip install chromadb && chroma run --path ./chroma_db
 *
 * then point CHROMA_URL at it (defaults to http://localhost:8000).
 *
 * We compute embeddings ourselves (Gemini) and hand Chroma precomputed vectors,
 * so no Chroma-side embedding function is involved.
 */
import crypto from 'node:crypto';
import { ChromaClient } from 'chromadb';
import { embedDocuments, embedQuery } from './embeddings.js';

const CHROMA_URL = process.env.CHROMA_URL || 'http://localhost:8000';

// A trivial embedding function so getOrCreateCollection never tries to pull a
// default one over the network; we never actually call it (we pass embeddings).
const noopEmbeddingFunction = {
  generate: async (texts) => texts.map(() => []),
};

let client = null;
function getClient() {
  if (!client) client = new ChromaClient({ path: CHROMA_URL });
  return client;
}

/** Stable per-document collection name so re-uploads reuse the same index. */
export function makeCollectionName(filename, content) {
  const digest = crypto.createHash('sha1').update(content).digest('hex').slice(0, 12);
  const safe = filename.replace(/[^a-zA-Z0-9]/g, '_').slice(0, 40);
  return `doc_${safe}_${digest}`;
}

export class VectorStore {
  constructor(collection, name) {
    this.collection = collection;
    this.name = name;
  }

  /** Open (or create) the collection for a document. */
  static async open(collectionName) {
    const collection = await getClient().getOrCreateCollection({
      name: collectionName,
      metadata: { 'hnsw:space': 'cosine' },
      embeddingFunction: noopEmbeddingFunction,
    });
    return new VectorStore(collection, collectionName);
  }

  async hasData() {
    try {
      return (await this.collection.count()) > 0;
    } catch {
      return false;
    }
  }

  /** Embed and index a list of chunk texts with their metadatas. */
  async addChunks(chunks, metadatas, { onProgress } = {}) {
    if (!chunks.length) return;
    const vectors = await embedDocuments(chunks, { onProgress });
    const ids = chunks.map((_, i) => `chunk_${i}`);
    const metas = metadatas || chunks.map((_, i) => ({ chunk_index: i }));
    await this.collection.add({
      ids,
      documents: chunks,
      embeddings: vectors,
      metadatas: metas,
    });
  }

  /**
   * Retrieve the top-k chunks for a query. Returns
   * [{ text, score, metadata }], score = cosine similarity (higher is better).
   */
  async query(question, k = 4) {
    const queryVec = await embedQuery(question);
    const res = await this.collection.query({
      queryEmbeddings: [queryVec],
      nResults: k,
    });
    const docs = (res.documents && res.documents[0]) || [];
    const metas = (res.metadatas && res.metadatas[0]) || [];
    const dists = (res.distances && res.distances[0]) || [];

    return docs.map((text, i) => ({
      text,
      score: 1 - Number(dists[i] ?? 1), // cosine distance -> similarity
      metadata: metas[i] || {},
    }));
  }
}
